### PR TITLE
Initial CDSDashboard Integration

### LINKED ISSUE #340: Integrate CDS-Dashboards into Qhub-cloud

### PR DESCRIPTION
**Pull request by Adam-D-Lewis**
This PR includes initial integration of CDS Dashboards into Qhub.  There are several limitations of the integration currently documented https://github.com/Quansight/qhub-cloud/issues/387.

This PR also includes some typo fixes in the docs.

Resolves #340

### CODE REVIEW

File: qhub/template/{{ cookiecutter.repo_directory }}/image/jupyterhub/environment.yaml [hunk 1]
```diff
@@ -9,4 +9,5 @@ dependencies:
  - escapism
  - pip:
    - jupyterhub-kubespawner >= 0.16.1
-   - qhub-jupyterhub-theme
+   - https://github.com/Quansight/qhub-jupyterhub-theme/archive/b78749053d5ddd094c684cf48a0b4e227880ea65.tar.gz
```
Comment (costrouc) [inline qhub/template/{{ cookiecutter.repo_directory }}/image/jupyterhub/environment.yaml:12]: I'd prefer if we cut releases for these and pinned them.

**Review by costrouc — CHANGES_REQUESTED** [review]

File: qhub/template/{{ cookiecutter.repo_directory }}/image/jupyterhub/environment.yaml (hunk 1 above)
Comment (costrouc) [reply qhub/template/{{ cookiecutter.repo_directory }}/image/jupyterhub/environment.yaml:12]: I'll create a release of qhub-jupyterhub-theme. Is it too early to have a cdsdashboard release?

Comment (danlester) [reply qhub/template/{{ cookiecutter.repo_directory }}/image/jupyterhub/environment.yaml:12]: It needs more testing for a full public release, but I can do a beta release on PyPI if that makes you more comfortable?

Comment (danlester) [reply qhub/template/{{ cookiecutter.repo_directory }}/image/jupyterhub/environment.yaml:12]: Should use cdsdashboards==0.5.0-b1. I can update both when you've done a qhub theme release (or feel free to do it yourself if quicker for you).

Comment (danlester) [reply qhub/template/{{ cookiecutter.repo_directory }}/image/jupyterhub/environment.yaml]: Now pinned to new versions

**Comment by costrouc** [comment]
Looks like this PR has merge conflicts

**Comment by costrouc** [comment]
@danlester this is a spotty action that fails sometimes https://github.com/Quansight/qhub-cloud/pull/388/checks?check_run_id=2175280125#step:18:12 not he fault of this PR... need to figure out how to fix this bug in another PR. Will review today.

**Comment by costrouc** [comment]
I will deploy this locally for testing before approving and merging but the code review look good to me.

**Comment by costrouc** [comment]
Not seeing the dashboard icon

```
Loading /etc/jupyterhub/config/values.yaml                                                                                                                                                                                         │
│ Loading /etc/jupyterhub/secret/values.yaml                                                                                                                                                                                         │
│ Loading extra config: jupyterhub_config                                                                                                                                                                                            │
│ [I 2021-03-24 15:01:10.961 JupyterHub app:2349] Running JupyterHub version 1.3.0                                                                                                                                                   │
│ [I 2021-03-24 15:01:10.961 JupyterHub app:2380] Using Authenticator: builtins.QHubPasswordAuthenticator                                                                                                                            │
│ [I 2021-03-24 15:01:10.961 JupyterHub app:2380] Using Spawner: cdsdashboards.hubextension.spawners.variablekube.VariableKubeSpawner-0.4.3                                                                                          │
│ [I 2021-03-24 15:01:10.961 JupyterHub app:2380] Using Proxy: jupyterhub.proxy.ConfigurableHTTPProxy-1.3.0                                                                                                                          │
│ [I 2021-03-24 15:01:10.961 JupyterHub app:1420] Loading cookie_secret from /srv/jupyterhub/jupyterhub_cookie_secret                                                                                                                │
│ [W 2021-03-24 15:01:10.973 JupyterHub app:1650] JupyterHub.hub_connect_port is deprecated as of 0.9. Use JupyterHub.hub_connect_url to fully specify the URL for connecting to the Hub.                                            │
│ [W 2021-03-24 15:01:10.974 JupyterHub auth:161] QHubPasswordAuthenticator.whitelist is deprecated in JupyterHub 1.2, use QHubPasswordAuthenticator.allowed_users instead                                                           │
│ [W 2021-03-24 15:01:10.974 JupyterHub app:1695] No admin users, admin interface will be unavailable.                                                                                                                               │
│ [W 2021-03-24 15:01:10.974 JupyterHub app:1697] Add any administrative users to `c.Authenticator.admin_users` in config.                                                                                                           │
│ [I 2021-03-24 15:01:11.039 JupyterHub reflector:280] watching for pods with label selector='component=singleuser-server' in namespace dev                                                                                          │
│ [I 2021-03-24 15:01:11.049 JupyterHub reflector:280] watching for events with field selector='involvedObject.kind=Pod' in namespace dev                                                                                            │
│ [I 2021-03-24 15:01:11.053 JupyterHub app:2077] example-user still running                                                                                                                                                         │
│ [I 2021-03-24 15:01:11.055 JupyterHub app:2419] Initialized 1 spawners in 0.028 seconds                                                                                                                                            │
│ [I 2021-03-24 15:01:11.056 JupyterHub app:2628] Not starting proxy                                                                                                                                                                 │
│ [I 2021-03-24 15:01:11.058 JupyterHub app:2664] Hub API listening on http://0.0.0.0:8081/hub/                                                                                                                                      │
│ [I 2021-03-24 15:01:11.058 JupyterHub app:2666] Private Hub API connect url http://10.99.253.50:8081/hub/                                                                                                                          │
│ [I 2021-03-24 15:01:11.058 JupyterHub app:2679] Starting managed service cull-idle                                                                                                                                                 │
│ [I 2021-03-24 15:01:11.058 JupyterHub service:339] Starting service 'cull-idle': ['python3', '/etc/jupyterhub/cull_idle_servers.py', '--url=http://127.0.0.1:8081/hub/api', '--timeout=1800', '--cull-every=600', '--concurrency=1 │
│ [I 2021-03-24 15:01:11.059 JupyterHub service:121] Spawning python3 /etc/jupyterhub/cull_idle_servers.py --url=http://127.0.0.1:8081/hub/api --timeout=1800 --cull-every=600 --concurrency=10                                      │
│ [I 2021-03-24 15:01:11.062 JupyterHub app:2688] Adding external service dask-gateway at http://traefik-dask-gateway.dev                                                                                                            │
│ [I 2021-03-24 15:01:11.066 JupyterHub proxy:319] Checking routes                                                                                                                                                                   │
│ [I 2021-03-24 15:01:11.066 JupyterHub app:2739] JupyterHub is now running at http://10.99.220.199:443/                                                                                                                             │
│ [I 2021-03-24 15:01:11.126 JupyterHub log:181] 200 GET /hub/api/users (cull-idle@127.0.0.1) 7.06ms                       
```

image

**Comment by costrouc** [comment]
My configuration

```yaml
  extraConfig:
    jupyterhub_config: |2



      # ================ CDSDASHBOARDS =====================
      c.JupyterHub.allow_named_servers = True
      from cdsdashboards.hubextension import cds_extra_handlers
      c.JupyterHub.extra_handlers = cds_extra_handlers
      from cdsdashboards.app import CDS_TEMPLATE_PATHS
      c.JupyterHub.template_paths = CDS_TEMPLATE_PATHS
      c.JupyterHub.spawner_class = 'cdsdashboards.hubextension.spawners.variablekube.VariableKubeSpawner'
      c.CDSDashboardsConfig.builder_class = 'cdsdashboards.builder.kubebuilder.KubeBuilder'
      c.VariableMixin.default_presentation_cmd = ['python3', '-m', 'jhsingle_native_proxy.main']

      c.JupyterHub.default_url = '/hub/home'



      # ==================== THEME =========================
      import tornado.web
      import qhub_jupyterhub_theme

      c.JupyterHub.extra_handlers += [
          (r'/custom/(.*)', tornado.web.StaticFileHandler, {"path": qhub_jupyterhub_theme.STATIC_PATH}),
      ]

      c.JupyterHub.template_paths = [
          qhub_jupyterhub_theme.TEMPLATE_PATH
      ] + c.JupyterHub.template_paths

      c.JupyterHub.template_vars = {
          "hub_title": "QHub - thisisatest",
          "hub_subtitle": "Autoscaling Compute Environment",
          "welcome": "Welcome to github-actions.qhub.dev. It is maintained by <a href=\"http://quansight.com\">Quansight staff</a>. The hub's configuration is stored in a github repository based on <a href=\"https://github.com/Quansight/qhub/\">https://github.com/Quansight/qhub/</a>. To provide feedback and report any technical problems, please use the <a href=\"https://github.com/Quansight/qhub/issues\">github issue tracker</a>.",
          "logo": "/hub/custom/images/jupyter_qhub_logo.svg",
          "primary_color": "#4f4173",
          "secondary_color": "#957da6",
          "accent_color": "#32C574",
          "text_color": "#111111",
          "h1_color": "#652e8e",
          "h2_color": "#652e8e",
      "cdsdashboards_enabled": True
      }

      # ================= Profiles =====================
      QHUB_USER_MAPPING = {'example-user': {'uid': 1000, 'primary_group': 'users', 'secondary_groups': ['admin'], 'password': '$2b$12$EvHy0mx3IxPiyZ5NWwBe.uTgayKPwkdA0WS0oJddbK8cW1WitpTwG'}}
      QHUB_GROUP_MAPPING = {'users': {'gid': 100}, 'admin': {'gid': 101}}
      QHUB_PROFILES = [{'display_name': 'Small Instance', 'description': 'Stable environment with 1 cpu / 1 GB ram', 'default': True, 'kubespawner_override': {'cpu_limit': 1, 'cpu_guarantee': 1, 'mem_limit': '1G', 'mem_guarantee': '1G', 'image': 'quansight/qhub-jupyterlab:1ece8e33b4b957c843c6f63150f0507bab60c422'}}, {'display_name': 'Medium Instance', 'description': 'Stable environment with 1.5 cpu / 2 GB ram', 'kubespawner_override': {'cpu_limit': 1.5, 'cpu_guarantee': 1.25, 'mem_limit': '2G', 'mem_guarantee': '2G', 'image': 'quansight/qhub-jupyterlab:1ece8e33b4b957c843c6f63150f0507bab60c422'}}]


      def qhub_generate_nss_files():
          passwd = []
          passwd_format = '{username}:x:{uid}:{gid}:{username}:/home/jovyan:/bin/bash'
          for username, config in QHUB_USER_MAPPING.items():
              uid = config['uid']
              gid = QHUB_GROUP_MAPPING[config['primary_group']]['gid']
              passwd.append(passwd_format.format(username=username, uid=uid, gid=gid))

          group = []
          group_format = '{groupname}:x:{gid}:'
          for groupname, config in QHUB_GROUP_MAPPING.items():
              gid = config['gid']
              group.append(group_format.format(groupname=groupname, gid=gid))

          return '\n'.join(passwd), '\n'.join(group)


      def qhub_list_admins(users):
          return [k for k,v in users.items() if v['primary_group'] == 'admin']


      def qhub_list_users(users):
          return [k for k,v in users.items() if v['primary_group'] != 'admin']


      def qhub_list_user_groups(username):
          user = QHUB_USER_MAPPING[username]
          return set([user['primary_group']] + user.get('secondary_groups', []))


      def qhub_configure_profile(username, safe_username, profile):
          user = QHUB_USER_MAPPING[username]
          uid = user['uid']
          primary_gid = QHUB_GROUP_MAPPING[user['primary_group']]['gid']
          secondary_gids = [QHUB_GROUP_MAPPING[_]['gid'] for _ in user.get('secondary_groups', [])]

          profile['kubespawner_override']['environment'] = {
             'LD_PRELOAD': 'libnss_wrapper.so',
             'NSS_WRAPPER_PASSWD': '/tmp/passwd',
             'NSS_WRAPPER_GROUP': '/tmp/group',
             'HOME': '/home/jovyan',
          }

          passwd, group = qhub_generate_nss_files()
          profile['kubespawner_override']['lifecycle_hooks'] = {
              "postStart": {
                  "exec": {
                      "command": ["/bin/sh", "-c", (
                           "echo '{passwd}' > /tmp/passwd && "
                           "echo '{group}' > /tmp/group && "
                           "ln -sfn /home/shared /home/jovyan/shared"
                      ).format(passwd=passwd, group=group)]
                  }
              }
          }

          profile['kubespawner_override']['init_containers'] = [
              {
                   "name": "init-nfs",
                   "image": "busybox:1.31",
                   "command": ["sh", "-c", ' && '.join([
                        "mkdir -p /mnt/home/{username}",
                        "chmod 700 /mnt/home/{username}",
                        "chown {uid}:{primary_gid} /mnt/home/{username}",
                        "mkdir -p /mnt/home/shared",
                        "chmod 777 /mnt/home/shared"
                   ] + ["mkdir -p /mnt/home/shared/{groupname} && chmod 770 /mnt/home/shared/{groupname} && chown 0:{gid} /mnt/home/shared/{groupname}".format(groupname=groupname, gid=config['gid']) for groupname, config in QHUB_GROUP_MAPPING.items()]).format(username=safe_username, uid=uid, primary_gid=primary_gid)],
                   "securityContext": {"runAsUser": 0},
                   "volumeMounts": [{"mountPath": "/mnt", "name": "home"}]
              }
          ]

          profile['kubespawner_override']['uid'] = uid
          profile['kubespawner_override']['gid'] = primary_gid
          profile['kubespawner_override']['supplemental_gids'] = secondary_gids
          profile['kubespawner_override']['fs_gid'] = primary_gid
          return profile

      def qhub_list_available_profiles(username):
          import escapism
          import string
          safe_chars = set(string.ascii_lowercase + string.digits)
          safe_username = escapism.escape(username, safe=safe_chars, escape_char='-').lower()

          exclude_keys = {'users', 'groups'}

          groups = qhub_list_user_groups(username)

          available_profiles = []
          for profile in QHUB_PROFILES:
              filtered_profile = qhub_configure_profile(username, safe_username, {k: v for k,v in profile.items() if k not in exclude_keys})

              if 'users' in profile:
                  if username in profile['users']:
                      available_profiles.append(filtered_profile)
              elif 'groups' in profile:
                  if len(groups & set(profile['groups'])) != 0:
                      available_profiles.append(filtered_profile)
              else:
                  available_profiles.append(filtered_profile)

          return available_profiles


      c.JupyterHub.admin_access = True
      c.Authenticator.admin_users = qhub_list_admins(QHUB_USER_MAPPING)
      c.Authenticator.whitelist = qhub_list_users(QHUB_USER_MAPPING)


      # ==================== AUTHENTICATION ======================
      # create passwords via
      # >>> bcrypt.hashpw(b'<password>', bcrypt.gensalt())
      import bcrypt
      from jupyterhub.auth import Authenticator

      class QHubPasswordAuthenticator(Authenticator):
          async def authenticate(self, handler, data):
              username, password = data['username'], data['password'].encode('utf-8')
              if username not in QHUB_USER_MAPPING:
                  return None

              if 'password' not in QHUB_USER_MAPPING[username]:
                  return None

              salted_password = QHUB_USER_MAPPING[username]['password'].encode('utf-8')
              if bcrypt.checkpw(password, salted_password):
                  return username
              return None

      c.JupyterHub.authenticator_class = QHubPasswordAuthenticator
  extraContainers: []

```

**Comment by danlester** [comment]
What happens if you go to /hub/dashboards?

My guess is that you are using an old qhub-jupyterhub-theme that doesn't explicitly expose the Dashboards menu.

In general, we perhaps need to check the Docker image tags that are in use right now for testing (and distribution), or let me know if there is a system.

I didn't change any Docker tags but made changes to the cookiecutter images assuming that somehow they would be built and find their way in!

**Comment by costrouc** [comment]
Will soon push resolutions to merge conflits. But I tested with newer docker images and it looks like it is working!

image

image

There were some gotcha with the dashboard showing up. But in general worked great especially for an initial release. Thanks for the great work @Adam-D-Lewis and @danlester  :smile:

**Review by costrouc — APPROVED** [review]

**Pull request merged**